libs/wire-api/src/Wire/API/Error.hs

```diff
@@ -0,0 +1,287 @@
+
+  route _ = route (Proxy @api)
+  hoistServerWithContext _ = hoistServerWithContext (Proxy @api)
+
+instance
+  (HasSwagger api, IsSwaggerError e) =>
+  HasSwagger (CanThrow e :> api)
+  where
+  toSwagger _ = addToSwagger @e (toSwagger (Proxy @api))
+
+instance HasSwagger api => HasSwagger (CanThrowMany '() :> api) where
+  toSwagger _ = toSwagger (Proxy @api)
+
+instance
+  (HasSwagger (CanThrowMany es :> api), IsSwaggerError e) =>
+  HasSwagger (CanThrowMany '(e, es) :> api)
+  where
+  toSwagger _ = addToSwagger @e (toSwagger (Proxy @(CanThrowMany es :> api)))
+
+type family DeclaredErrorEffects api :: EffectRow where
+  DeclaredErrorEffects (CanThrow e :> api) = (ErrorEffect e ': DeclaredErrorEffects api)
+  DeclaredErrorEffects (CanThrowMany '(e, es) :> api) =
+    DeclaredErrorEffects (CanThrow e :> CanThrowMany es :> api)
+  DeclaredErrorEffects (x :> api) = DeclaredErrorEffects api
+  DeclaredErrorEffects (Named n api) = DeclaredErrorEffects api
+  DeclaredErrorEffects api = '[]
+
+errorResponseSwagger :: forall e. KnownError e => S.Response
+errorResponseSwagger =
+  mempty
+    & S.description .~ (eMessage err <> " (label: `" <> eLabel err <> "`)")
+    & S.schema ?~ S.Inline (S.toSchema (Proxy @(SStaticError e)))
+  where
+    err = dynError @e
+
+addStaticErrorToSwagger :: forall e. KnownError e => S.Swagger -> S.Swagger
+addStaticErrorToSwagger =
+  S.allOperations . S.responses . S.responses . at (fromIntegral (eCode err))
+    %~ Just . addRef
+  where
+    err = dynError @e
+    resp = errorResponseSwagger @e
+
+    addRef :: Maybe (S.Referenced S.Response) -> S.Referenced S.Response
+    addRef Nothing = S.Inline resp
+    addRef (Just (S.Inline resp1)) = S.Inline (combineResponseSwagger resp1 resp)
+    addRef (Just r@(S.Ref _)) = r
+
+type family MapError (e :: k) :: StaticError
+
+type family ErrorEffect (e :: k) :: Effect
+
+class IsSwaggerError e where
+  addToSwagger :: S.Swagger -> S.Swagger
+
+-- | An effect for a static error type with no data.
+type ErrorS e = Error (Tagged e ())
+
+throwS :: forall e r a. (Member (ErrorS e) r) => Sem r a
+throwS = throw (Tagged @e ())
+
+noteS :: forall e r a. (Member (ErrorS e) r) => Maybe a -> Sem r a
+noteS = note (Tagged @e ())
+
+mapErrorS ::
+  forall e e' r a.
+  (Member (ErrorS e') r) =>
+  Sem (ErrorS e ': r) a ->
+  Sem r a
+mapErrorS = mapError (Tagged @e' . unTagged)
+
+mapToRuntimeError ::
+  forall e e' r a. Member (Error e') r => e' -> Sem (ErrorS e ': r) a -> Sem r a
+mapToRuntimeError e' = mapError (const e')
+
+mapToDynamicError ::
+  forall e r a.
+  (Member (Error DynError) r, KnownError (MapError e)) =>
+  Sem (ErrorS e ': r) a ->
+  Sem r a
+mapToDynamicError = mapToRuntimeError (dynError @(MapError e))
+
+errorToWai :: forall e. KnownError (MapError e) => Wai.Error
+errorToWai = toWai (dynError @(MapError e))
+
+runErrorS :: forall e a r. Sem (ErrorS e ': r) a -> Sem r (Maybe a)
+runErrorS = fmap hush . runError
```

Comment:
Why is it OK to hush a statically known error?

Reply:
Nevermind, this is not used. I removed it and cleaned up the module a bit.